engine/crates/gateway-core/src/lib.rs

```diff
@@ -124,53 +141,28 @@ where
-                    vec![engine::ServerError::new("Unauthorized", None)],
-                    // doesn't really matter, this is not client facing
-                    OperationType::Query,
-                )),
-                Default::default(),
-            ));
-        };
-
-        if !self.cache_config.global_enabled || !self.cache_config.partial_registry.enable_caching {
-            let response = Arc::clone(&self.executor)
-                .execute(Arc::clone(ctx), auth, request)
-                .await?;
-
-            return Ok((Arc::new(response), Default::default()));
-        }
-
-        match build_cache_key(&self.cache_config, ctx.as_ref(), &request, &auth) {
-            Ok(cache_key) => {
-                let execution_fut = Arc::clone(&self.executor)
-                    .execute(Arc::clone(ctx), auth, request)
-                    .instrument(info_span!("execute"))
-                    .map(|res| res.map(Arc::new));
-
-                let cached_execution =
-                    cache::cached_execution(&self.cache, cache_key, ctx.as_ref(), execution_fut).await;
-
-                cache::process_execution_response(ctx.as_ref(), cached_execution)
-            }
-            Err(_) => {
-                let result = Arc::clone(&self.executor)
-                    .execute(Arc::clone(ctx), auth, request)
-                    .instrument(info_span!("execute"))
-                    .map(|res| res.map(Arc::new))
-                    .await?;
-
-                let response = CachedExecutionResponse::Origin {
-                    response: result,
-                    cache_read: CacheReadStatus::Bypass,
-                };
-
-                cache::process_execution_response(ctx.as_ref(), Ok(response))
-            }
+        let normalized_query = operation_normalizer::normalize(request.query(), request.operation_name()).ok();
+        let (response, headers) = self.execute_with_auth(ctx, request, auth).await?;
+        if let Some((operation, normalized_query)) = response.graphql_operation.clone().zip(normalized_query) {
+            self.operation_metrics.record(
+                grafbase_tracing::metrics::GraphqlOperationMetricsAttributes {
+                    ty: match operation.r#type {
+                        common_types::OperationType::Query { .. } => "query",
+                        common_types::OperationType::Mutation => "mutation",
+                        common_types::OperationType::Subscription => "subscription",
+                    },
+                    name: operation.name,
+                    normalized_query_hash: blake3::hash(normalized_query.as_bytes()).into(),
+                    normalized_query,
+                    has_errors: !response.errors.is_empty(),
+                    cache_status: headers
+                        .get(X_GRAFBASE_CACHE)
+                        .and_then(|v| v.to_str().ok().map(|s| s.to_string())),
+                },
+                start.elapsed(),
```

Comment:
I assume I'm missing something: why are we doing this here _and_ in the engine?

Reply:
because gateway-core deals with the caching so need to be here, but the stream interface doesn't let me retrieve the necessary information. So need to do in the engine for stream

Reply:
The alternative would be imposing constraints on `Executor::StreamingResponse` to be able to retrieve headers and send all the metadata through it

Reply:
> So need to do in the engine for stream

Ah, that'll be what I'm missing - didn't notice it was stream only